File: app/src/test/java/org/mifos/mobilebanking/LoanAccountDetailPresenterTest.java
package org.mifos.mobilebanking;

import android.content.Context;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.models.accounts.loan.LoanWithAssociations;
import org.mifos.mobilebanking.presenters.LoanAccountsDetailPresenter;
import org.mifos.mobilebanking.ui.views.LoanAccountsDetailView;
import org.mifos.mobilebanking.util.RxSchedulersOverrideRule;
import org.mifos.mobilebanking.utils.Constants;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import io.reactivex.Observable;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner.class)
public class LoanAccountDetailPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    LoanAccountsDetailView view;

    private LoanAccountsDetailPresenter presenter;
    private LoanWithAssociations loanWithAssociations;

    @Before
    public void setUp() throws Exception {
        presenter = new LoanAccountsDetailPresenter(dataManager, context);
        presenter.attachView(view);

        loanWithAssociations = FakeRemoteDataSource.getLoanAccountWithTransaction();
    }

    @Test
    public void testLoadLoanAccountDetails() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 4)).
                thenReturn(Observable.just(loanWithAssociations));

        presenter.loadLoanAccountDetails(4);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showLoanAccountsDetail(loanWithAssociations);
        verify(view, never()).showErrorFetchingLoanAccountsDetail(context.
                getString(R.string.error_loan_account_details_loading));
    }

    @Test
    public void testLoadLoanAccountDetailsFails() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 4)).
                thenReturn(Observable.<LoanWithAssociations>error(new RuntimeException()));

        presenter.loadLoanAccountDetails(4);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showErrorFetchingLoanAccountsDetail(context.
                getString(R.string.error_loan_account_details_loading));
        verify(view, never()).showLoanAccountsDetail(null);
    }
}

File: app/src/test/java/org/mifos/mobile/LoanAccountDetailPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.accounts.loan.LoanWithAssociations
import org.mifos.mobile.presenters.LoanAccountsDetailPresenter
import org.mifos.mobile.ui.views.LoanAccountsDetailView
import org.mifos.mobile.util.RxSchedulersOverrideRule
import org.mifos.mobile.utils.Constants

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner::class)
class LoanAccountDetailPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: LoanAccountsDetailView? = null
    private var presenter: LoanAccountsDetailPresenter? = null
    private var loanWithAssociations: LoanWithAssociations? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = LoanAccountsDetailPresenter(dataManager!!, context)
        presenter?.attachView(view)
        loanWithAssociations = FakeRemoteDataSource.loanAccountWithTransaction
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetails() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 4)).thenReturn(Observable.just(loanWithAssociations))
        presenter?.loadLoanAccountDetails(4)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showLoanAccountsDetail(loanWithAssociations)
        Mockito.verify(view, Mockito.never())?.showErrorFetchingLoanAccountsDetail(context?.getString(R.string.error_loan_account_details_loading))
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetailsFails() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 4)).thenReturn(Observable.error(RuntimeException()))
        presenter?.loadLoanAccountDetails(4)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showErrorFetchingLoanAccountsDetail(context?.getString(R.string.error_loan_account_details_loading))
        Mockito.verify(view, Mockito.never())?.showLoanAccountsDetail(null)
    }
}